src/server/req_preemptjob.c

```diff
@@ -126,112 +357,54 @@ req_preemptjobs(struct batch_request *preq)
  *
  * @param[in] code - determines if the job was preempted or not.
  * @param[in] aux  - determines the method by which job was preempted.
- * @param[in] local_preq - contains the job information and the scheuler
- *                         request to which reply is to be sent.
+ * @param[in] pjob - the job in which we are replying to the preemption request
  */
 
 void
-reply_preempt_jobs_request(int code, int aux, struct batch_request *local_preq)
+reply_preempt_jobs_request(int code, int aux, struct job *pjob)
 {
-	job *pjob;
-	int newsub;
-	int newstate;
-	attribute temphold;
-	struct batch_request *preq = local_preq->rq_nest;
-	struct batch_reply *preply = &(preq->rq_reply);
-
-	static int count = 0;
-	static int index = 0;
-	static preempt_job_info	*preempt_jobs_list = NULL;
-
-	count = preq->rq_ind.rq_preempt.count;
-	pjob = local_preq->rq_extra;
+	struct batch_request *preq;
 
-	if (preempt_jobs_list == NULL) {
-		if ((preempt_jobs_list = calloc(sizeof(preempt_job_info), count)) == NULL) {
-			log_err(errno, __func__, "Unable to allocate memory");
-			return;
-		}
-	}
+	if (pjob == NULL)
+		return;
 
-	preply->brp_choice = BATCH_REPLY_CHOICE_PreemptJobs;
-	preply->brp_un.brp_preempt_jobs.count = preq->rq_ind.rq_preempt.count;
+	preq = pjob->ji_pmt_preq;
 
 	if (code != PBSE_NONE) {
-		switch ((int)pjob->preempt_order[0].order[pjob->preempt_order_index+1]) {
-			case PREEMPT_METHOD_SUSPEND:
-				issue_signal(pjob, SIG_SUSPEND, post_signal_req, pjob, preq);
-				break;
-			case PREEMPT_METHOD_CHECKPOINT:
-				shutdown_preempt_chkpt(pjob, preq);
-				break;
-			case PREEMPT_METHOD_REQUEUE:
-				issue_signal(pjob, SIG_RERUN, post_rerun, pjob, preq);
-				break;
-			default:
-				/* preemption failed */
-				preply->brp_code = 1;
-				strcpy(preempt_jobs_list[index].order, "000");
-				sprintf(preempt_jobs_list[index].job_id, "%s", pjob->ji_qs.ji_jobid);
-				index++;
-				break;
+		if (pjob->preempt_order[0].order[pjob->preempt_order_index] == PREEMPT_METHOD_CHECKPOINT)
+			clear_preempt_hold(pjob);
+		if (issue_preempt_request((int)pjob->preempt_order[0].order[pjob->preempt_order_index + 1], pjob, preq)) {
```

Comment:
The reason I did it this way is control will fall down to where I check if we're done and send the preempt reply.  This might be the final job, and we need to send the ack back to the scheduler.  If we made an else and returned from there, I'd have to move the check to reply in there as well as the bottom of the function.  I think it is good as it is.  Let me know what you think.

Reply:
You are right, I misread the code :)